src/writer.rs

```diff
@@ -510,6 +510,38 @@ impl Writer {
         Ok(())
     }
 
+    /// Write a function that associates codepoints with other codepoints.
+    ///
+    /// The function will use a match expression to map between codepoints.
+    /// The fallback branch of the match returns 0.
```

Comment:
Hmm. I wonder if this is the right choice since `0` is itself a valid codepoint, although it is often used as a sentinel in practice. So maybe it's fine. Certainly, it would be much nicer to return an `Option`, but I'm not sure what kind of impact that would have on code size/performance. I wonder if perhaps `Option<std::num::NonZeroU32>` would be a better choice here? You could then return an error at codegen time if any of the destination codepoints are equivalent to `0`. You'd then need to use `NonZeroU32::new_unchecked`.

I'm not sure whether this is getting too cute though. Using an `Option` kind of assumes that all uses of this generated function will branch on the result value to see if the mapping was successful. But if there's a use case where one doesn't need that branch, then the `Option` just kind of gets in the way.

Reply:
Valid concerns. I did ponder this at the time. Ultimately I settled on 0 because:

* It was simple
* None of the bidi mirroring glyphs map to 0, although since this method could be used for something else `NUL` being a valid codepoint could be a problem in the future.
* For our use case in Prince that motivated this code, the previous C code I was rewriting used 0 as the fallback value.
* In our specific use case as it stands now if we move to `Option`, the Rust code in our FFI would need to map `None` to 0 anyway, and then the other side of the FFI will also check for 0, so it adds a redundant branch for every codepoint we process. I am fully aware this is a general tool though, so I'm fine with it diverging a little from our ideal case.

I did some experiments with an `Option<std::num::NonZeroU32>` implementation:

* Compile time didn't seem noticeable different (although I did not measure this).
* Code size is identical (as expected):

```
File .text   Size                Crate Name
0.1%  1.7% 2.4KiB bidi_mirroring_glyph bidi_mirroring_glyph::non_zero::bidi_mirroring_glyph_option
0.1%  1.7% 2.4KiB bidi_mirroring_glyph bidi_mirroring_glyph::match_ch
```

* Performance on a benchmark that applies the function to every codepoint and unwraps the value returned, shows only a slight difference over the existing implementation:

```
Bidi_Mirror/match/paragraph
                        time:   [3.3374 us 3.3384 us 3.3395 us]

Bidi_Mirror/match non-zero/paragraph
                        time:   [3.6334 us 3.6350 us 3.6365 us]
```

I'll tidy the code up and update the PR for you check it out.